### PR TITLE
[pt2] meta for `searchsorted.Scalar`, tests, and out support

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #106283

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -10927,7 +10927,11 @@
   dispatch:
     CPU: searchsorted_cpu
     CUDA: searchsorted_cuda
-  autogen: searchsorted.Scalar_out
+
+- func: searchsorted.Scalar_out(Tensor sorted_sequence, Scalar self, *, bool out_int32=False, bool right=False, str? side=None, Tensor? sorter=None, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
+    CPU: searchsorted_out_cpu
+    CUDA: searchsorted_out_cuda
```
Comment (nkaretnikov) [inline aten/src/ATen/native/native_functions.yaml:10934]: I'll take a closer look at this. Looks like autogen doesn't generate the out variant. Maybe I'll fix the codegen instead.

```py
In [42]: torch.searchsorted(xs, torch.tensor(2), out=torch.tensor((), dtype=torch.int64))
Out[42]: tensor(6)

In [43]: torch.searchsorted(xs, 2, out=torch.tensor((), dtype=torch.int64))
TypeError: searchsorted() received an invalid combination of arguments - got (Tensor, int, out=Tensor), but expected one of:
 * (Tensor sorted_sequence, Tensor input, *, bool out_int32, bool right, str side, Tensor sorter, Tensor out)
 * (Tensor sorted_sequence, Number self, *, bool out_int32, bool right, str side, Tensor sorter)
```

File: test/functorch/test_vmap.py [hunk 2]
```diff
@@ -3740,6 +3740,7 @@ def test_vmap_exhaustive(self, device, dtype, op):
         xfail('ge', device_type='cuda'),
         xfail('_upsample_bilinear2d_aa'),
         xfail('argsort'),  # aten::argsort.stable hit the vmap fallback which is currently disabled
+        xfail('searchsorted'),  # aten::searchsorted.Scalar hit the vmap fallback which is currently disabled
```
Comment (nkaretnikov) [inline test/functorch/test_vmap.py:3743]: Scalar wasn't tested in OpInfos before.

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -5139,13 +5139,16 @@ def t_(self):
...
 @out_wrapper()
 def meta_searchsorted(
     sorted_sequence, self, *, out_int32=False, right=False, side=None, sorter=None
 ):
     dtype = torch.int32 if out_int32 else torch.int64
-    return torch.empty_like(self, dtype=dtype).contiguous()
+    if isinstance(self, torch.Tensor):
+        return torch.empty_like(self, dtype=dtype).contiguous()
+    else:  # Scalar
+        return torch.empty((), dtype=dtype, device=sorted_sequence.device)
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:5151]: Meta now supports Scalar.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -2984,34 +2984,40 @@ def error_inputs_bucketize(opinfo, device, **kwargs):
...
-        ((M, M), ((M, M),)),
-        ((0, 0, 0), ((0, 0, 0),)),
-        ((M, M, M), ((M, M, M),)),
+        ((0,), ((0,),), False),
+        ((M,), ((), (M,), (M, M)), False),
+        ((0, 0), ((0, 0),), False),
+        ((M, M), ((M, M),), False),
+        ((0, 0, 0), ((0, 0, 0),), False),
+        ((M, M, M), ((M, M, M),), False),
+        ((L,), ((),), True),
```
Comment (nkaretnikov) [inline torch/testing/_internal/common_methods_invocations.py:2998]: Added one more case for testing Scalar.

**Comment by nkaretnikov** [comment]
From the list in https://github.com/pytorch/pytorch/issues/105105

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (nkaretnikov) [reply aten/src/ATen/native/native_functions.yaml:10934]: I've looked at this a bit. It seems it's a known issue with autogen: it does generate things on the aten side, but not on the Python side.

```py
    # A list of operators that are expected to be auto-generated for this NativeFunction.
    # Note: This list isn't actually directly used by the codegen to generate anything.
    # Instead, the codegen figures out what operators to generate purely based off of
    # function schema, and uses the autogen declarations to error check.
    # We expect every NativeFunction that gets auto-generated be explicitly called out
    # in native_functions.yaml
    autogen: List["OperatorName"]
```

So it seems like a serious undertaking to fix this on the codegen side. I propose we just add out variants to unblock this.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**